crypto/src/lib.rs

```diff
@@ -384,52 +506,69 @@ impl PublicKey {
 #[cfg(not(feature = "ffi_import"))]
 impl From<PrivateKey> for PublicKey {
     fn from(private_key: PrivateKey) -> Self {
-        Self::try_from_private(private_key).expect("can't fail for valid `PrivateKey`")
+        let digest_function = private_key.digest_function();
+        let key_gen_option = Some(KeyGenOption::FromPrivateKey(private_key));
+
+        match digest_function {
+            Algorithm::Ed25519 => {
+                PublicKey::Ed25519(ed25519::Ed25519Sha512::keypair(key_gen_option).0)
+            }
+            Algorithm::Secp256k1 => {
+                PublicKey::Secp256k1(secp256k1::EcdsaSecp256k1Sha256::keypair(key_gen_option).0)
+            }
+            Algorithm::BlsNormal => PublicKey::BlsNormal(bls::BlsNormal::keypair(key_gen_option).0),
+            Algorithm::BlsSmall => {
+                PublicKey::BlsSmall(Box::new(bls::BlsSmall::keypair(key_gen_option).0))
+            }
+        }
     }
 }
 
 ffi::ffi_item! {
     /// Private Key used in signatures.
-    #[derive(Clone, PartialEq, Eq, CopyGetters)]
-    #[cfg_attr(not(feature="ffi_import"), derive(DebugCustom, Display, Serialize))]
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "hex::encode_upper(payload)"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "hex::encode_upper(payload)"))]
-    pub struct PrivateKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        #[serde(with = "hex::serde")]
-        payload: ConstVec<u8>,
+    #[derive(Clone, PartialEq, Eq)]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PrivateKey {
+        Ed25519(Box<ed25519::PrivateKey>),
+        Secp256k1(secp256k1::PrivateKey),
+        BlsNormal(bls::PrivateKey),
+        BlsSmall(bls::PrivateKey),
     }
 }
 
 #[ffi_impl_opaque]
 impl PrivateKey {
     /// Key payload
-    // TODO: Derive with getset once FFI impl is fixed
-    pub fn payload(&self) -> &[u8] {
-        self.payload.as_ref()
+    pub fn payload(&self) -> ConstVec<u8> {
+        match self {
+            Self::Ed25519(key) => key.to_keypair_bytes().to_vec().into(),
```

Comment:
```suggestion
            Self::Ed25519(key) => key.to_keypair_bytes().into(),
```

we should have a shortcut like this instead of creating intermediate `Vec`

Reply:
Created `ToConstVecExt` trait